### PR TITLE
Feature/file type support SDK-3856 SDK-3844

### PR DESCRIPTION
**Pull request by piyush-kukadiya**
- Adds implementation for file pre loading of any file type.
- Adds public API for clearing files from cache and disk
- Adds public API for returning absolute file path for given file url 
- Adds public API to check if file for given url exists or not

### CODE REVIEW

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java [hunk 1]
```diff
@@ -3515,35 +3510,83 @@ public void clearInAppResources(boolean expiredOnly) {
...
+        }
+    }
+
+    /**
+     * Checks if a file exists for the given non-null url.
+     *
+     * @param url the non-null url for which to check the existence of the file.
+     * @return true if a file exists for the specified url, false otherwise.
+     */
+    public boolean isFileExistsForUrl(@NonNull String url){
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java:3570]: this can be renamed to `checkFileForUrl` or `doesFileExistForUrl`

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt [hunk 2]
```diff
@@ -4,83 +4,97 @@ import android.content.Context
...
-            null
-        }
-    },
-    private val fileToBytes: (file: File?) -> ByteArray? = { file ->
-        file?.readBytes()
-    },
     private val inAppRemoteSource: InAppImageFetchApiContract = InAppImageFetchApi()
 ) {
+
+    private var ctCaches: CTCaches = CTCaches.instance(
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt:28]: names to call args and formatting

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt [hunk 3]
```diff
@@ -4,83 +4,97 @@ import android.content.Context
...
+        return isFileCached(url)
     }
 
-    fun isGifCached(url: String) : Boolean {
-        val gifMemoryCache = ctCaches.gifCache()
+    fun isFileCached(url: String): Boolean {
+        val memoryAccessObjectList = listOf<MemoryAccessObject<*>>(
+            FileMemoryAccessObject(ctCaches), ImageMemoryAccessObject(ctCaches),
+            GifMemoryAccessObject(ctCaches)
+        )
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt:80]: can we avoid the * wildcard?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt [hunk 4]
```diff
@@ -4,83 +4,97 @@ import android.content.Context
...
+        memoryAccessObjectList.forEach {
+            val file = it.fetchDiskMemory(url)
+            if (file != null) {
+                return true
+            }
+        }
+        return false
     }
 
     fun cachedImage(cacheKey: String?): Bitmap? {
```
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt:100]: Can we rename to getCachedImage. Would be more readable?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/ImageMemoryV1.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import kotlin.math.max
+
+class ImageMemoryV1(
+    private val config: MemoryConfig,
+    private val logger: ILogger? = null
+) : Memory<Bitmap> {
+
+    private var imageInMemory: LruCache<Pair<Bitmap, File>>? = null
+    private var imageDiskMemory: FileCache? = null
+    private val lock1 = Any()
```
Comment (Anush-Shand) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/ImageMemoryV1.kt:17]: Names for the lock can be more specific

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepo.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    fun fetchAllImages(urls: List<String>)
+    fun fetchAllGifs(urls: List<String>)
+    fun fetchAllFiles(
+        urls: List<String>,
+        completionCallback: (status: Boolean, urlStatusMap: Map<String, Boolean>) -> Unit
+    )
+
+    fun cleanupStaleImages(validUrls: List<String>)
+    fun cleanupStaleFiles(validUrls: List<String>)
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepo.kt:19]: cleanupStaleGifs seems missing, intentional?

**Comment by Anush-Shand** [comment]
@piyush-kukadiya GifMemoryV1, ImageMemoryV1 and FileMemoryV2 have same implementation for same functions, can they be implemented in the base class itself?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java (hunk 1 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java]: Fixed in [49f583](https://github.com/CleverTap/clevertap-android-sdk/commit/49f583b5f6f025a8ca3d1f0a70a612b3166f0822)

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt (hunk 2 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt:28]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/24ae903b1e807dc71671dda11097b5cbb99dacf4

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java [hunk 7]
```diff
@@ -3515,35 +3510,83 @@ public void clearInAppResources(boolean expiredOnly) {
...
+    /**
+     * Deletes all types of files which are preloaded for SDK features like custom in-app templates, app functions and
+     * variables etc.
+     *
+     * @param expiredOnly to clear only files which will not be needed further for SDK features like custom in-app
+     *                    templates, app functions and variables etc.
+     */
+    public void clearFileResources(boolean expiredOnly) {
+
+        Logger logger = coreState.getConfig().getLogger();
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/a786506e881526fa7aca2efe7a12c6cb859e22c8 https://github.com/CleverTap/clevertap-android-sdk/commit/b52be621b792329a790bf53c1c0b52bb0205897a https://github.com/CleverTap/clevertap-android-sdk/commit/0442041e5c7e308720c750e467e0fafbe1c0f680

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java [hunk 8]
```diff
@@ -3515,35 +3510,83 @@ public void clearInAppResources(boolean expiredOnly) {
...
+    }
+
+    /**
+     * Checks if a file exists for the given non-null url.
+     *
+     * @param url the non-null url for which to check the existence of the file.
+     * @return true if a file exists for the specified url, false otherwise.
+     */
+    public boolean isFileExistsForUrl(@NonNull String url){
+        Logger logger = coreState.getConfig().getLogger();
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/b52be621b792329a790bf53c1c0b52bb0205897a

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/ImageMemoryAccessObject.kt [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.File
+
+class ImageMemoryAccessObject(private val ctCaches: CTCaches) : MemoryAccessObject<Bitmap> {
+
+    override fun fetchInMemory(key: String): Pair<Bitmap, File>? {
+        val imageInMemory = ctCaches.imageCache()
+        return imageInMemory.get(key)
+    }
+
+    override fun fetchInMemoryAndTransform(key: String, transformTo: Int): Any? {
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/ImageMemoryAccessObject.kt]: > can we have specific return types and not `Any?`

Right now return type Any produces 6 methods. Changing Any to specific return type will produce 18 methods(12 more).

[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/ImageMemoryAccessObject.kt]: > `transformTo: Int` can be an enum/sealed class?

fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/2bbb5f6593e354a558e0fb9e203ea670f5e867ea

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepo.kt (hunk 6 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepo.kt:19]: It's intentional because `cleanupStaleImages` clears both gif and images.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapFactory.java [hunk 10]
```diff
@@ -188,6 +193,7 @@ public Void call() throws Exception {
                 false,
                 storeRegistry,
                 triggersManager,
+                templatesManager,
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapFactory.java:196]: this is not part of this pr

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/InAppController.java [hunk 11]
```diff
@@ -851,7 +857,23 @@ private static void showInApp(Context context, final CTInAppNotification inAppNo
...
+                        config.getLogger().verbose(config.getAccountId(),
+                                "file download status from showInApp() = " + aBoolean + " and url status map = "
+                                        + stringBooleanMap);
+                        if (aBoolean)
+                        {
+                            inAppController.presentTemplate(inAppNotification);
+                        }
+                        return null;
+                    });
+                }
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/InAppController.java:876]: we should check if file required to present this template has finished download. 

Will this block wait for all resources to finish downloading?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/FileResourceProvider.kt [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+import java.io.ByteArrayOutputStream
+import java.io.File
+
+internal class FileResourceProvider(
+    private val images: File,
+    private val gifs: File,
+    private val allFileTypesDir: File,
+    private val logger: ILogger? = null,
+    private val inAppRemoteSource: FileFetchApiContract = FileFetchApi()
+) {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/FileResourceProvider.kt:25]: this is a feature specific class and it should only have following methods 

saveFile()
hasFile()/isFIleCached()
getFile(cacheOnly: boolean)

This class can encapsulate another sub feature class which was `InAppResourceProvider`

Can we not have so many methods in this class as it is confusing to understand.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/FileMemoryAccessObject.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            when(transformTo)
+            {
+                MEMORY_DATA_TRANSFORM_TO_BITMAP -> bytesToBitmap(it.first)
+                MEMORY_DATA_TRANSFORM_TO_BYTEARRAY -> it.first
+                MEMORY_DATA_TRANSFORM_TO_FILE -> it.second
+            }
+        }
+    }
+
+    override fun fetchDiskMemoryAndTransform(key: String, transformTo: MemoryDataTransformationType): Any? {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/FileMemoryAccessObject.kt:28]: can we have specific return type

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryAccessObject.kt [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    BitmapFactory.decodeByteArray(
+        it,
+        0,
+        it.size
+    )
+}
+val bitmapToBytes: (bitmap: Bitmap?) -> ByteArray? = {
+    it?.let {
+        val stream = ByteArrayOutputStream()
+        it.compress(Bitmap.CompressFormat.PNG, 100, stream)
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryAccessObject.kt:35]: what was the quality in past, do we keep it as 100?

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryCreator.kt [hunk 15]
```diff
@@ -0,0 +1,64 @@
+package com.clevertap.android.sdk.inapp.images.memory
+
+import android.graphics.Bitmap
+import com.clevertap.android.sdk.ILogger
+import java.io.File
+
+class MemoryCreator {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryCreator.kt:7]: this can be an object.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepoFactory.kt [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+import android.content.Context
+import com.clevertap.android.sdk.Logger
+import com.clevertap.android.sdk.inapp.images.FileResourceProvider
+import com.clevertap.android.sdk.inapp.images.cleanup.FileCleanupStrategy
+import com.clevertap.android.sdk.inapp.images.cleanup.FileCleanupStrategyExecutors
+import com.clevertap.android.sdk.inapp.images.preload.FilePreloaderExecutors
+import com.clevertap.android.sdk.inapp.images.preload.FilePreloaderStrategy
+import com.clevertap.android.sdk.inapp.store.preference.StoreRegistry
+
+internal class FileResourcesRepoFactory {
```
Comment (CTLalit) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepoFactory.kt:12]: this can be an object

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNotification.java [hunk 17]
```diff
@@ -452,33 +458,54 @@ boolean isTablet() {
...
-                } else {
-                    this.error = "Error processing image as bitmap was NULL";
+            int index = 0;
+            while (index < fileUrlMetas.size()) {
+                Pair<String, CtCacheType> data = fileUrlMetas.get(index);
+                byte[] bytes = fileResourceProvider.fetchFile(data.getFirst());
+
+                if (bytes == null) {
+                    // download fail
+                    this.error = "some error";
```
Comment (vasct) [inline clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNotification.java:475]: We need to discuss how to handle failed files with product. We could leave failed files as "null" and still display the in-app with no error. 

Meanwhile could we change the error string to something more descriptive. "Failed to download file" or something.

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/repo/InAppImageRepoTest.kt [hunk 18]
```diff
@@ -10,36 +11,38 @@ import io.mockk.verify
 import org.junit.Test
 
 class InAppImageRepoTest {
-
-    private val inAppImageCleanupStrategy = mockk<InAppCleanupStrategy>(relaxed = true)
-    private val preloaderStrategy = mockk<InAppImagePreloaderStrategy>(relaxed = true)
+/*
```
Comment (vasct) [inline clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/repo/InAppImageRepoTest.kt:14]: Same

File: clevertap-core/src/test/java/com/clevertap/android/sdk/utils/CTCachesTest.kt [hunk 19]
```diff
@@ -2,12 +2,14 @@ package com.clevertap.android.sdk.utils
 
 import com.clevertap.android.sdk.TestLogger
 import org.junit.After
+import org.junit.Ignore
 import org.junit.Test
 import kotlin.test.assertEquals
 
+@Ignore
```
Comment (vasct) [inline clevertap-core/src/test/java/com/clevertap/android/sdk/utils/CTCachesTest.kt:9]: Why Ignore

**Review by vasct — COMMENTED** [review]
Good work overall. Why are so many tests commented-out ?

**Review by CTLalit — APPROVED** [review]
Approved with open details in https://docs.google.com/spreadsheets/d/17BBJ0NCOVc7UMEaISSkFjGzEEx_kslHFM1mUOBUn0OI/edit?gid=0#gid=0

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt (hunk 3 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt]: its needed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/InAppResourceProviderTest.kt [hunk 20]
```diff
@@ -5,40 +5,48 @@ import com.clevertap.android.sdk.TestLogger
...
 import kotlin.test.assertEquals
 import kotlin.test.assertNotNull
 
+@Ignore
 class InAppResourceProviderTest {
 
     private val mockCache = Mockito.mock(CTCaches::class.java)
-    private val mockBitmap = Mockito.mock(Bitmap::class.java)
+    private val mockBitmapPair = Pair(Mockito.mock(Bitmap::class.java), mockk<File>())
+    private val mockBytesPair = Pair(Mockito.mock(ByteArray::class.java), mockk<File>())
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/InAppResourceProviderTest.kt:22]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/5305e4c1666d3ef2fac96a4c517e039964a0a861

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt (hunk 4 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/InAppResourceProvider.kt]: there is no equivalent method that starts with prefix `set` let's keep as it is. https://stackoverflow.com/a/10568713

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/InAppController.java (hunk 11 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/InAppController.java]: resolved

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/FileResourceProvider.kt (hunk 12 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/FileResourceProvider.kt:30]: resolved

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/FileMemoryAccessObject.kt (hunk 13 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/FileMemoryAccessObject.kt]: requires generics to reduce duplicate code and functions

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryAccessObject.kt (hunk 14 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryAccessObject.kt:35]: It's taken from Jan release code of image downloading.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryCreator.kt (hunk 15 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/memory/MemoryCreator.kt:7]: Idea was to create static kind of methods equivalent to factory pattern and not singleton. Kotlin docs https://kotlinlang.org/docs/object-declarations.html#companion-objects and other blogs shows companion object way of creating factories. It's still debatable keeping as it is as of now.

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepoFactory.kt (hunk 16 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/images/repo/FileResourcesRepoFactory.kt:12]: Same as https://github.com/CleverTap/clevertap-android-sdk/pull/619#discussion_r1676035340

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNotification.java (hunk 17 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppNotification.java]: Error is fixed

File: clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java [hunk 21]
```diff
@@ -3515,39 +3510,87 @@ public String getLocale() {
...
-                preloadStrategy,
-                inAppAssetStore,
-                legacyInAppStore
-        );
         if (expiredOnly) {
-            impl.cleanupStaleImagesNow();
+            impl.cleanupStaleFiles();
         } else {
-            impl.cleanupAllImages();
+            impl.cleanupStaleFiles(); // todo this also only clears expired ones. fixme
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/CleverTapAPI.java]: fixed already

File: clevertap-core/src/main/java/com/clevertap/android/sdk/bitmap/HttpBitmapLoader.kt
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/bitmap/HttpBitmapLoader.kt:1]: Idea was to use it as an adapter. `FileFetchAPI#makeApiCallForFile` will act as an adapter.

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/ImpressionManagerTest.kt [hunk 22]
```diff
@@ -18,7 +18,7 @@ import kotlin.test.assertEquals
 
 class ImpressionManagerTest : BaseTestCase() {
 
-    @Mock
+    /*@Mock
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/ImpressionManagerTest.kt]: fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplateInAppDataTest.kt [hunk 23]
```diff
@@ -1,12 +1,35 @@
...
+import com.clevertap.android.sdk.inapp.customtemplates.CustomTemplateContext.TemplateContext
+import io.mockk.*
 import org.json.JSONObject
 import org.junit.*
+import org.mockito.Mockito.*
 import kotlin.test.assertEquals
 
 class CustomTemplateInAppDataTest {
 
+/*
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplateInAppDataTest.kt]: already fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/FileResourceProviderTest.kt [hunk 24]
```diff
@@ -0,0 +1,261 @@
+package com.clevertap.android.sdk.inapp.images
+
+import android.graphics.Bitmap
+import com.clevertap.android.sdk.TestLogger
+import io.mockk.*
+import org.junit.Test
+import java.io.File
+
+class FileResourceProviderTest {
+/*
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/FileResourceProviderTest.kt]: already fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/cleanup/FileCleanupStrategyCoroutineTest.kt [hunk 25]
```diff
@@ -1,43 +1,43 @@
...
     private val dispatchers = TestDispatchers(testScheduler)
 
-    private val cleanupStrategy = InAppCleanupStrategyCoroutine(
-        inAppResourceProvider = inAppResourceProvider,
+    private val cleanupStrategy = FileCleanupStrategyCoroutine(
+        fileResourceProvider = mFileResourceProvider,
         dispatchers = dispatchers
     )
-
+/*
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/cleanup/FileCleanupStrategyCoroutineTest.kt]: already fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/cleanup/FileCleanupStrategyExecutorsTest.kt [hunk 26]
```diff
@@ -1,21 +1,21 @@
...
+    private val mFileResourceProvider = mockk<FileResourceProvider>(relaxed = true)
     private val executors = MockCTExecutors()
 
-    private val cleanupStrategy = InAppCleanupStrategyExecutors(
-        inAppResourceProvider = inAppResourceProvider,
+    private val cleanupStrategy = FileCleanupStrategyExecutors(
+        fileResourceProvider = mFileResourceProvider,
         executor = executors
-    )
+    )/*
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/cleanup/FileCleanupStrategyExecutorsTest.kt]: already fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/preload/InAppImagePreloaderExecutorsTest.kt [hunk 27]
```diff
@@ -16,16 +16,16 @@ class InAppImagePreloaderExecutorsTest {
...
     private val logger = TestLogger()
 
-    private val inAppImagePreloader = InAppImagePreloaderExecutors(
-        inAppImageProvider = inAppResourceProvider,
+    private val inAppImagePreloader = FilePreloaderExecutors(
+        fileResourceProvider = mFileResourceProvider,
         logger = logger,
         executor = executors
-    )
+    )/*
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/preload/InAppImagePreloaderExecutorsTest.kt]: already fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/repo/InAppImageRepoTest.kt (hunk 18 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/inapp/images/repo/InAppImageRepoTest.kt]: already fixed

File: clevertap-core/src/test/java/com/clevertap/android/sdk/utils/CTCachesTest.kt (hunk 19 above)
Comment (piyush-kukadiya) [reply clevertap-core/src/test/java/com/clevertap/android/sdk/utils/CTCachesTest.kt:9]: Fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/d5c573fd09510eedd1d9321a561611e15ac7a569

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplateContext.kt [hunk 28]
```diff
@@ -253,8 +268,8 @@ sealed class CustomTemplateContext private constructor(
                         else -> overrides.getDouble(argument.name)
                     }
                 }
-                //TODO CustomTemplates add FILE handling when implemented
-                FILE -> null
+                // TODO CustomTemplates add FILE handling when implemented
```
[the remark this answers is not in the document]
Comment (piyush-kukadiya) [reply clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplateContext.kt:272]: fixed in https://github.com/CleverTap/clevertap-android-sdk/commit/fd32a93e32618610b93dcc3b9441eb16be83167a
